Sources/xcodeproj/Objects/Project/PBXProjEncoder.swift

```diff
@@ -1,5 +1,45 @@
 import Foundation
 
+// Defines the sorting applied to files within the file lists. Defaults to by UUID.
+public enum FileListSortOrder {
+    case byUUID
+    case byFilename
+    
+    func sort<Object>(lhs: (PBXObjectReference, Object), rhs: (PBXObjectReference, Object)) -> Bool where Object: PlistSerializable & Equatable {
+        return lhs.0 < rhs.0
+    }
+    
+    func sort(lhs: (PBXObjectReference, PBXBuildFile), rhs: (PBXObjectReference, PBXBuildFile)) -> Bool {
+        switch self {
+        case .byFilename:
+            return lhs.1.file?.path ?? lhs.1.uuid < rhs.1.file?.path ?? rhs.1.uuid
+        default:
+            return lhs.0 < rhs.0
+        }
+    }
+    
+    func sort(lhs: (PBXObjectReference, PBXFileReference), rhs: (PBXObjectReference, PBXFileReference)) -> Bool {
+        switch self {
+        case .byFilename:
+            return lhs.1.path ?? lhs.1.uuid < rhs.1.path ?? rhs.1.uuid
+        default:
+            return lhs.0 < rhs.0
+        }
+    }
+}
+
+// Defines the sorting applied to groups with the project navigator and various build phases.
+public enum ProjectNavigatorSortOrder {
+    case unsorted
+    case byFilename
+    case byFilenameGroupsFirst
```

Comment:
I like how you defined a function, `sort` in the `FileListSortOrder` type that you pass to the Foundation `sort` method for sorting. Is there any reason why you haven't followed a similar approach for `ProjectNavigatorSortOrder`, `BuildPhaseFileSortOrder`. If not, I'd define it here because we can easily test it without having to test the whole encoder.

Reply:
Have moved the other sort closures into the enums as well.